src/github-handler/commit-and-push-handler.ts

```diff
@@ -0,0 +1,198 @@
+ * @param {string} refHead the base of the new commit(s)
+ * @param {string} treeSha the tree SHA that this commit will point to
+ * @param {string} message the message of the new commit
+ * @returns {Promise<string>} the new commit SHA
+ */
+async function createCommit(
+  logger: Logger,
+  octokit: Octokit,
+  origin: RepoDomain,
+  refHead: string,
+  treeSha: string,
+  message: string
+): Promise<string> {
+  const commitData = (
+    await octokit.git.createCommit({
+      owner: origin.owner,
+      repo: origin.repo,
+      message,
+      tree: treeSha,
+      parents: [refHead],
+    })
+  ).data;
+  logger.info(`Successfully created commit. See commit at ${commitData.url}`);
+  return commitData.sha;
+}
+
+/**
+ * Update a reference to a SHA
+ * Rejects if GitHub V3 API fails with the GitHub error response
+ * @param {Logger} logger The logger instance
+ * @param {Octokit} octokit The authenticated octokit instance
+ * @param {RepoDomain} origin the the remote repository to push changes to
+ * @param {string} refName the name of the branch ref
+ * @param {string} newSha the ref to update the commit HEAD to
+ * @returns {Promise<void>}
+ */
+async function updateRef(
+  logger: Logger,
+  octokit: Octokit,
+  origin: RepoDomain,
+  refName: string,
+  newSha: string
+): Promise<void> {
+  await octokit.git.updateRef({
+    owner: origin.owner,
+    repo: origin.repo,
+    ref: `heads/${refName}`,
+    sha: newSha,
+  });
+  logger.info(`Successfully updated reference ${refName} to ${newSha}`);
+}
+
+/**
+ * Given a set of changes, apply the commit(s) on top of the given branch's head and upload it to GitHub
+ * Rejects if GitHub V3 API fails with the GitHub error response
+ * @param {Logger} logger The logger instance
+ * @param {Octokit} octokit The authenticated octokit instance
+ * @param {string} refHead the base of the new commit(s)
+ * @param {Changes} changes the set of repository changes
+ * @param {RepoDomain} origin the the remote repository to push changes to
+ * @param {string} originBranchName the remote branch that will contain the new changes
+ * @param {string} commitMessage the message of the new commit
+ * @returns {Promise<void>}
+ */
+async function commitAndPush(
+  logger: Logger,
+  octokit: Octokit,
+  refHead: string,
+  changes: Changes,
+  origin: RepoDomain,
+  originBranchName: string,
+  commitMessage: string
+) {
+  try {
+    const tree = generateTreeObjects(changes);
+    const treeSha = await createTree(logger, octokit, origin, refHead, tree);
+    const commitSha = await createCommit(
+      logger,
+      octokit,
+      origin,
+      refHead,
+      treeSha,
+      commitMessage
+    );
+    await updateRef(logger, octokit, origin, originBranchName, commitSha);
+  } catch (err) {
+    logger.error('Error while creating a tree and updating the ref');
+    throw err;
+  }
+}
+
+export {
+  commitAndPush,
+  createCommit,
+  generateTreeObjects,
+  createTree,
+  updateRef,
+};
```

Comment:
To consider: let's think about what we truly want/need to export (we may need to for testing individual functions)